File: medium/Hash-Table/intersection-of-two-arrays/intersection.py
from typing import List

class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        """
        Find unique intersection elements of two arrays.
        Approach: Hash Set for fast lookup
        Time Complexity: O(m + n)
        Space Complexity: O(m)
        """
        set1 = set(nums1)
        res = []
        
        for num in nums2:
            if num in set1:
                res.append(num)
                set1.remove(num)  # avoid duplicate
        
        return res

File: medium/Hash-Table/intersection-of-two-arrays/intersection.cpp
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        /*
         * Get unique intersection of two arrays.
         * Approach: Hash Set lookup
         * Time: O(m + n)
         * Space: O(m)
         */
        unordered_set<int> st(nums1.begin(), nums1.end());
        vector<int> res;
        
        for(int x : nums2){
            if(st.count(x)){
                res.push_back(x);
                st.erase(x);
            }
        }
        return res;
    }
};
